src/Mapping/DataTarget/Direct.php

```diff
@@ -137,13 +138,49 @@ protected function checkAssignData($newData, $valueContainer, $getter)
         $currentData = $valueContainer->$getter($this->language);
         DataObject::setHideUnpublished($hideUnpublished);
 
-        if (!empty($currentData) && $this->writeIfTargetIsNotEmpty === false) {
+        $fieldName = lcfirst(str_replace('get', '', $getter));
+        $fieldDefinition = $this->getFieldDefinition($valueContainer, $fieldName);
+        if ($this->writeIfTargetIsNotEmpty === false && !$fieldDefinition->isEmpty($currentData)) {
             return false;
         }
-        if ($this->writeIfSourceIsEmpty === false && (empty($newData) || ($newData instanceof QuantityValue && empty($newData->getValue())))) {
+
+        if ($this->writeIfSourceIsEmpty === false && $fieldDefinition->isEmpty($newData)) {
             return false;
         }
 
         return true;
     }
+
+    /**
+     * @param DataObject\Concrete|DataObject\Objectbrick\Data\AbstractData $valueContainer
+     * @param string $fieldName
```

Comment:
@dvesh3 Why do we have docs there, when we can do this in the parameter types?

Reply:
doesn't work with typehint as `checkAssignData` defines `Object` for `$valueContainer` so better introduce it with major version change the typehint for `checkAssignData` as well (bc break).